src/is_subscribed.rs

```diff
@@ -22,70 +21,62 @@ fn is_subscribed(
 }
 
+/// user has subscribed and confirmed subscription by clicking on the double opt-in email
 fn check_if_subscribed(topic: Option<String>, user_data: serde_json::Value) -> bool {
+    let sub = user_data
+        .as_object()
+        .and_then(|v| v.get("subscription"))
+        .and_then(|v| v.as_object());
+
     let subscribed = if let Some(topic) = topic {
-        user_data
-            .as_object()
-            .and_then(|v| v.get("subscription"))
-            .and_then(|v| v.as_object())
-            .and_then(|v| v.get("topics"))
+        sub.and_then(|v| v.get("topics"))
             .and_then(|v| v.as_array())
             .map(|topics| topics.iter().any(|t| t.as_str() == Some(&topic)))
             .unwrap_or_default()
     } else {
-        user_data
-            .as_object()
-            .and_then(|v| v.get("subscription"))
-            .and_then(|v| v.as_object())
-            .and_then(|v| v.get("subscribed"))
+        sub.and_then(|v| v.get("subscribed"))
             .and_then(|v| v.as_bool())
             .unwrap_or_default()
     };
 
-    subscribed
+    let confirmed = sub
+        .and_then(|v| v.get("confirmed"))
+        .and_then(|v| v.as_bool())
+        .unwrap_or_default();
+
+    subscribed && confirmed
 }
 
-fn user_data_from_sid_or_tid(
+/// try to get user data from the session's uid or session's data->'subscription_uid'
+fn user_data_from_sid(
     ft_sdk::Cookie(user_id_from_session): ft_sdk::Cookie<{ ft_sdk::auth::SESSION_KEY }>,
-    ft_sdk::Cookie(user_id_from_tracker): ft_sdk::Cookie<"fastn-tid">,
     conn: &mut ft_sdk::Connection,
 ) -> Result<serde_json::Value, ft_sdk::Error> {
     use diesel::prelude::*;
 
-    let query;
-    if user_id_from_session.is_some() {
+    let query = if user_id_from_session.is_some() {
         let user_id = user_id_from_session.unwrap();
-        query = diesel::sql_query(
+        diesel::sql_query(
             r#"
             SELECT
                 fastn_user.data as data
             FROM fastn_user
             JOIN fastn_session
+            ON fastn_user.id = fastn_session.uid 
+            OR fastn_user.id = json_extract(fastn_session.data, '$.subscription_uid')
             WHERE
                 fastn_session.id = $1
-                AND fastn_user.id = fastn_session.uid
-            LIMIT 1;
-            "#,
-        )
-        .bind::<diesel::sql_types::Text, _>(user_id);
-    } else if user_id_from_tracker.is_some() {
-        let user_id = user_id_from_tracker.unwrap();
-        query = diesel::sql_query(
-            r#"
-            SELECT
-                fastn_user.data as data
-            FROM fastn_user
-            JOIN fastn_tracker
-            WHERE
-                fastn_tracker.id = $1
-                AND fastn_user.id = fastn_tracker.uid
+                AND (
+                    fastn_user.id = fastn_session.uid
+                    OR fastn_user.id = json_extract(fastn_session.data, '$.subscription_uid')
+                )
```

Comment:
Why do we have repeat this both in `WHERE` and `ON` clause?

Reply:
yea this was redundant, thanks for pointing out. Removed in https://github.com/fastn-community/subscription/pull/2/commits/cc03497e2d446f4903ea129287db0f80a1a286a9